platform/core/src/services/HangingProtocolService/HangingProtocolService.ts

```diff
@@ -649,38 +698,181 @@ class HangingProtocolService {
+
+    let firstNotDisabled: number;
+
+    for (let i = 0; i < stages.length; i++) {
+      if (stages[i].enable === 'enabled') return i;
+      if (firstNotDisabled === undefined && stages[i].enable !== 'disabled') {
+        firstNotDisabled = i;
+      }
+    }
+
+    return firstNotDisabled;
+  }
+
+  /** Gets the current state, used to record the state when trying to apply items */
+  public getCurrentState(): Record<string, unknown> {
+    return {
+      protocol: this.protocol,
+      stage: this.stage,
+      viewportMatchDetails: this.viewportMatchDetails,
+      displaySetMatchDetails: this.displaySetMatchDetails,
+      activeImageLoadStrategyName: this.activeImageLoadStrategyName,
+    };
+  }
+
+  private _setProtocol(
+    protocol: HangingProtocol.Protocol,
+    options = null as HangingProtocol.SetProtocolOptions
+  ): void {
+    const old = this.getCurrentState();
+
     try {
+      if (!this.protocol || this.protocol.id !== protocol.id) {
+        this.stage = options?.stageIndex || 0;
+        this.protocol = this._copyProtocol(protocol);
+
+        const { imageLoadStrategy } = protocol;
+        if (imageLoadStrategy) {
+          // check if the imageLoadStrategy is a valid strategy
+          if (
+            this.registeredImageLoadStrategies[imageLoadStrategy] instanceof
+            Function
+          ) {
+            this.activeImageLoadStrategyName = imageLoadStrategy;
+          }
+        }
+
+        this._updateStageActivation(options);
+      }
+
+      const stage = this._findStage(options);
+      if (stage === undefined) {
+        throw new Error(
+          `Can't find applicable stage ${protocol.id} ${options?.stageIndex}`
+        );
+      }
+      this.stage = stage as number;
       this._updateViewports(options);
     } catch (error) {
-      this.protocol = oldProtocol;
+      console.log(error);
+      Object.assign(this, old);
       throw new Error(error);
     }
 
-    this._broadcastChange(this.EVENTS.PROTOCOL_CHANGED, {
-      viewportMatchDetails: this.viewportMatchDetails,
-      displaySetMatchDetails: this.displaySetMatchDetails,
-      hpAlreadyApplied: this.hpAlreadyApplied,
-      protocol: this.protocol,
-    });
+    if (options?.restoreProtocol !== true) {
+      this._broadcastEvent(HangingProtocolService.EVENTS.PROTOCOL_CHANGED, {
+        viewportMatchDetails: this.viewportMatchDetails,
+        displaySetMatchDetails: this.displaySetMatchDetails,
+        protocol: this.protocol,
+        stageIdx: this.stage,
+        stage: this.protocol.stages[this.stage],
+        activeStudyUID: this.activeStudy?.StudyInstanceUID,
+      });
+    } else {
+      this._broadcastEvent(HangingProtocolService.EVENTS.RESTORE_PROTOCOL, {
+        protocol: this.protocol,
+        stageIdx: this.stage,
+        stage: this.protocol.stages[this.stage],
+        activeStudyUID: this.activeStudy?.StudyInstanceUID,
+      });
+    }
+  }
+
+  public getStageIndex(hpId: string, options): number {
```

Comment:
here you have hpId, and couple of lines below it is hangingProtocolId, I suggest sticking to the hangginProtocolId for consistency and reaadability